src/dune_rules/pkg_rules.ml

```diff
@@ -255,19 +255,32 @@ module Pkg = struct
     { id : Id.t
     ; build_command : Build_command.t option
     ; install_command : Dune_lang.Action.t option
-    ; depends : t list
+    ; depends : t list Memo.Lazy.t
+    ; post_depends : t list Memo.Lazy.t
     ; info : Pkg_info.t
     ; paths : Paths.t
     ; files_dir : Path.Build.t
     ; mutable exported_env : string Env_update.t list
     }
 
-  module Top_closure = Top_closure.Make (Id.Set) (Monad.Id)
+  module Top_closure = Top_closure.Make (Id.Set) (Memo)
+
+  let deps t =
+    let* depends = Memo.Lazy.force t.depends in
+    Memo.List.concat_map depends ~f:(fun dep ->
+      let+ post_depends =
+        Memo.Lazy.force dep.post_depends
+        (* It's possible for a package to be a post dependency of one of its
+           dependencies but we avoid tracking that dependency here to prevent
+           cycles in the dependency hierarchy. *)
+        >>| List.filter ~f:(fun dep -> not (Id.equal t.id dep.id))
```

Comment:
I still don't understand why this check is correct. If the intent is to make sure post deps never introduce cycles, this isn't enough as it only removes cycles of length 1.

Reply:
Hmm I need to think about this some more. I see that since your changes the tests are now failing with:
```
Error: Dependency cycle between:
     - package bar
  [1]
```
which is the type of memo cycle that this check was trying to prevent.

But as you point out, memo cycles may still happen if the loop would be longer than 1. I'll think about this some more today.